File: user-service/src/main/java/br/com/userservice/application/exception/ExceptionGlobalHandler.java
package br.com.userservice.application.exception;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.validation.BindException;
import org.springframework.validation.FieldError;
import org.springframework.validation.ObjectError;
import org.springframework.web.bind.MethodArgumentNotValidException;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.ResponseStatus;

import java.util.List;

@ControllerAdvice
public class ExceptionGlobalHandler {

    @ExceptionHandler(ValidationException.class)
    public ResponseEntity<?> handleValidationException(ValidationException validationException) {
        var details = new ExceptionDetails(HttpStatus.BAD_REQUEST.value(), validationException.getMessage());
        return new ResponseEntity<>(details, HttpStatus.BAD_REQUEST);
    }
    @ExceptionHandler(MethodArgumentNotValidException.class)
    public ResponseEntity<?> handleValidationException(MethodArgumentNotValidException exception) {
        var details = new ExceptionDetails(HttpStatus.BAD_REQUEST.value(), "Field " + exception.getFieldError().getField() + " invalid");
        return new ResponseEntity<>(details, HttpStatus.BAD_REQUEST);
    }

    @ExceptionHandler(BindException.class)
    public ResponseEntity<?> handleValidationException(BindException exception) {
        var details = new ExceptionDetails(HttpStatus.BAD_REQUEST.value(), "Field " + exception.getFieldError().getField() + " invalid");
        return new ResponseEntity<>(details, HttpStatus.BAD_REQUEST);
    }
}

File: validated-document-service/src/main/kotlin/br/com/validateddocumentservice/application/exception/ExceptionGlobalHandler.kt
package br.com.validateddocumentservice.application.exception

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler

@ControllerAdvice
class ExceptionGlobalHandler {
    @ExceptionHandler(ValidationException::class)
    fun handleValidationException(validationException: ValidationException): ResponseEntity<*>? {
        val details = ExceptionDetails(HttpStatus.BAD_REQUEST.value(), validationException.message)
        return ResponseEntity<Any>(details, HttpStatus.BAD_REQUEST)
    }
}